pennylane/tape/qscript.py

```diff
@@ -1410,3 +1411,76 @@ def wrapper(*args, **kwargs):
         return qscript
 
     return wrapper
+
+
+def bind_new_parameters_tape(
+    tape: QuantumScript, params: Sequence[TensorLike], indices: Sequence[int]
+):
+    """Create a new tape with updated parameters.
+
+    This function takes a :class:`~.tape.QuantumScript` as input, and returns
+    a new ``QuantumScript`` containing the new parameters at the provided indices,
+    with the parameters at all other indices remaining the same.
+
+    Args:
+        tape (.tape.QuantumScript): Tape to update
+        params (Sequence[TensorLike]): New parameters to create the tape with. This
+            must have the same length as ``indices``.
+        indices (Sequence[int]): The parameter indices to update with the given parameters.
+            The index of a parameter is defined as its index in ``tape.get_parameters()``.
+
+    Returns:
+        .tape.QuantumScript: New tape with updated parameters
+    """
+    # pylint: disable=no-member
+
+    if len(params) != len(indices):
+        raise ValueError("Number of provided parameters does not match number of indices")
+
+    new_ops = []
+    idx = 0
+    p_idx = 0
+
+    for op in tape.circuit:
+        if isinstance(op, Operator):
+            data = op.data
+        elif op.obs is not None:
+            data = op.obs.data
+        else:
+            data = ()
+
+        # determine if any parameters of the operator need to be rebinded
+        if any(i + idx in indices for i in range(len(data))):
```

Comment:
by only checking existence in `indices`, we don't care about order. I'm not _totally_ opposed to that, but perhaps we should make it clear. Example:
```pycon
>>> qs = qml.tape.QuantumScript([qml.RX(1.1, 0), qml.RY(2.2, 1)])
>>> qs.bind_new_parameters([3.3, 4.4], [0, 1]).circuit
[RX(3.3, wires=[0]), RY(4.4, wires=[1])]
>>> qs.bind_new_parameters([3.3, 4.4], [1, 0]).circuit
[RX(3.3, wires=[0]), RY(4.4, wires=[1])]
```
We could instead follow today's typical strategy of a `trainable_only` boolean toggle, setting indices to `tape.trainable_params if trainable_only else range(len(tape._par_info))`, since that seems to be the only 2 cases. That said, we'd lose some generality

Reply:
This behaviour is the same as `set_parameters`, see the test `test_setting_parameters_unordered` in `test_tape.py`.

I also considered a `trainable_only` flag, but as you said it loses some generality. One example is for the broadcasting transforms where the indices correspond to parameters containing a batch dimension.